daemon/containerd/image_builder.go

```diff
@@ -268,27 +316,45 @@ type rwlayer struct {
 	c           *containerd.Client
 	snapshotter string
 	root        string
+	lease       leases.Lease
 }
 
 func (rw *rwlayer) Root() string {
 	return rw.root
 }
 
-func (rw *rwlayer) Commit() (builder.ROLayer, error) {
-	// we need this here for the prepared snapshots or
-	// we'll have racy behaviour where sometimes they
-	// will get GC'd before we commit/use them
-	ctx, _, err := rw.c.WithLease(context.TODO(), leases.WithRandomID(), leases.WithExpiration(1*time.Hour))
-	if err != nil {
-		return nil, fmt.Errorf("failed to create lease for commit: %w", err)
-	}
+func (rw *rwlayer) Commit() (_ builder.ROLayer, outErr error) {
 	snapshotter := rw.c.SnapshotService(rw.snapshotter)
 
 	key := stringid.GenerateRandomID()
+
+	lm := rw.c.LeasesService()
+	ctx := context.TODO()
+	lease, err := leaseSnapshot(ctx, lm, rw.snapshotter, key)
+	if err != nil {
+		return nil, err
+	}
+
 	err = snapshotter.Commit(ctx, key, rw.key)
 	if err != nil && !cerrdefs.IsAlreadyExists(err) {
 		return nil, err
 	}
+	defer func() {
```

Comment:
Why didn't you stick with the `client.WithLease` that returns a cleanup function? Also I think you can only remove the lease and the snapshot will be GC'd automatically for you

Reply:
Ah, right, removed the manual snapshot removal (which would be incorrect and cause used snapshot by other layer to be removed if the snapshot already exists, but later operation fails).

There are 2 separate leases created here:
1. Long lived (created by `leaseSnapshot`) that's removed in the `Release`.
2. Short lived - this one is only to hold the content created by `rootfs.CreateDiff` and released when this function exits and the resource is attached to the long lived lease.

The reason for having them both is because sometimes the content was GC'd before the `lm.AddResource` got called (which I believe was the reason to have the fixed 1h leases in the initial implementation). And we can't add the lease upfront like with the snapshot, because we don't know what digest the diff archive will have.

Reply:
That's true, fixed.